From 77bebbdae623249d9c5d0a6770a123f72ce6be9b Mon Sep 17 00:00:00 2001
From: AaeonCM <aaeon_cm@aaeon.com>
Date: Thu, 6 Jan 2022 11:36:17 +0000
Subject: [PATCH 2/4] GPIO: Add workaround to enable Open-Drain GPIO pins

The bwlow APL pins on UP-APL03 are open-drain circuits and
need disable "GPIOTxDis" in order to achieve high-Z pin state
(or external pull-up) for GPIO ouput high function
(Intel EDS #559360).
    LPSS_I2C0_SDA
    LPSS_I2C0_SCL,
    LPSS_I2C1_SDA
    LPSS_I2C1_SCL

Signed-off-by: AaeonCM <aaeon_cm@aaeon.com>
---
 drivers/pinctrl/intel/pinctrl-intel.c | 14 ++++++++++++++
 1 file changed, 14 insertions(+)

diff --git a/drivers/pinctrl/intel/pinctrl-intel.c b/drivers/pinctrl/intel/pinctrl-intel.c
index 812d3d48afb3..deb677f44c34 100644
--- a/drivers/pinctrl/intel/pinctrl-intel.c
+++ b/drivers/pinctrl/intel/pinctrl-intel.c
@@ -869,6 +869,11 @@ static int intel_gpio_get(struct gpio_chip *chip, unsigned int offset)
 	if (!(padcfg0 & PADCFG0_GPIOTXDIS))
 		return !!(padcfg0 & PADCFG0_GPIOTXSTATE);
 
+	/* Workaround to enable APL open-drain GPIO*/
+	if (!strcmp(chip->label, "INT3452:02") && offset >= 0 && offset < 4 &&
+		(padcfg0 & PADCFG0_GPIOTXDIS) && (padcfg0 & PADCFG0_GPIORXDIS))
+		return !!(padcfg0 & PADCFG0_GPIOTXSTATE);
+
 	return !!(padcfg0 & PADCFG0_GPIORXSTATE);
 }
 
@@ -895,6 +900,15 @@ static void intel_gpio_set(struct gpio_chip *chip, unsigned int offset,
 		padcfg0 |= PADCFG0_GPIOTXSTATE;
 	else
 		padcfg0 &= ~PADCFG0_GPIOTXSTATE;
+
+	/* Workaround to enable APL open-drain GPIO*/
+	if (!strcmp(chip->label, "INT3452:02") && offset >= 0 && offset < 4) {
+		if (value)
+			padcfg0 |= PADCFG0_GPIOTXDIS;
+		else
+			padcfg0 &= ~PADCFG0_GPIOTXDIS;
+	}
+
 	writel(padcfg0, reg);
 	raw_spin_unlock_irqrestore(&pctrl->lock, flags);
 }
-- 
2.17.1

